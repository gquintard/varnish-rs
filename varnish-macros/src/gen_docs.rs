//! Code to generate automated documentation into a file

use std::fmt::Write;
use std::fs;
use std::path::Path;

use crate::model::{FuncInfo, FuncType, ParamInfo, ParamKind, ParamType, ParamTypeInfo, VmodInfo};

// Small helpers to write to a string without checking the result

macro_rules! wrt {
    ($docs:expr, $($arg:tt)*) => {
        let _ = write!($docs, $($arg)*);
    };
}
macro_rules! ln {
    ($docs:expr, $($arg:tt)*) => {
        let _ = writeln!($docs, $($arg)*);
    };
}

/// Generate documentation for the VMOD and save it to a file
pub fn generate_docs(info: &VmodInfo) {
    let Some(ref doc_file) = info.params.docs else {
        return; // doc file is not set, skipping
    };
    let Ok(dir) = std::env::var("CARGO_MANIFEST_DIR") else {
        panic!("Unable to get the CARGO_MANIFEST_DIR env var to save documentation, you may need to remove the `docs` parameter from the `#[vmod]` attribute");
    };
    let doc_file = Path::new(&dir).join(doc_file);
    let docs = gen_doc_content(info);
    if let Err(e) = fs::write(doc_file.as_path(), docs) {
        panic!(
            "Unable to save documentation to file {}: {e}",
            doc_file.display()
        );
    }
}

/// Generate documentation for the VMOD as a single string
pub fn gen_doc_content(info: &VmodInfo) -> String {
    let mut docs = String::new();

    // Initial warning to not edit this file
    ln!(
        docs,
        r#"<!--

   !!!!!!  WARNING: DO NOT EDIT THIS FILE!

   This file was generated from the Varnish VMOD source code.
   It will be automatically updated on each build.

-->"#
    );

    // generate documentation
    ln!(docs, "# Varnish Module (VMOD) `{}`", info.ident);
    write_docs(&mut docs, &info.docs, "#");
    ln!(
        docs,
        r#"
```vcl
// Place import statement at the top of your VCL file
// This loads vmod from a standard location
import {ident};

// Or load vmod from a specific file
import {ident} from "path/to/lib{ident}.so";
```"#,
        ident = info.ident
    );

    for func in &info.funcs {
        if !matches!(func.func_type, FuncType::Function) {
            continue;
        }
        write_function(&mut docs, "###", "Function", func);
    }

    for obj in &info.objects {
        ln!(docs, "\n### Object `{}`", obj.ident);

        write_docs(&mut docs, &obj.docs, "###");
        write_function(&mut docs, "####", &obj.ident, &obj.constructor);
        for method in &obj.funcs {
            write_function(&mut docs, "####", "Method", method);
        }
    }

    docs
}

fn write_function(mut docs: &mut String, prefix: &str, obj_or_typ: &str, func: &FuncInfo) {
    let user_args = get_user_args(func);

    if matches!(func.func_type, FuncType::Constructor) {
        ln!(
            docs,
            r#"
```vcl
// Create a new instance of the object in your VCL init function
sub vcl_init {{
    new {ident} = {obj_or_typ}.{sig};
}}
```"#,
            ident = func.ident,
            sig = fn_sig(func, &user_args),
        );
    } else {
        wrt!(docs, "\n{prefix} {obj_or_typ} ");
        ln!(&mut docs, "{}", fn_sig(func, &user_args));
    }

    write_docs(docs, &func.docs, prefix);

    // List of arguments are printed only if any of them have documentation
    if user_args.iter().any(|(arg, _)| !arg.docs.is_empty()) {
        ln!(docs, "");
        for (arg, ty) in &user_args {
            wrt!(docs, "* `{}`:", bracketed_name(arg, ty));
            if arg.docs.is_empty() {
                ln!(docs, "");
            } else {
                write_docs(docs, &arg.docs, prefix);
            }
        }
    }
}

fn fn_sig(func: &FuncInfo, user_args: &Vec<(&ParamTypeInfo, &ParamInfo)>) -> String {
    let mut res = String::new();
    let is_md_txt = matches!(
        func.func_type,
        FuncType::Function | FuncType::Method | FuncType::Event
    );
    if is_md_txt {
        wrt!(res, "`");
    }
    if matches!(func.func_type, FuncType::Function | FuncType::Method) {
        wrt!(res, "{} ", func.returns.value_type().to_vcc_type());
    }
    wrt!(res, "{}(", func.ident);
    let mut first = true;
    for (arg, ty) in user_args {
        if first {
            first = false;
        } else {
            wrt!(res, ", ");
        }
        res.push_str(&bracketed_name(arg, ty));
        if !ty.default.is_null() {
            wrt!(res, " = {}", ty.default);
        }
    }
    wrt!(res, ")");
    if is_md_txt {
        wrt!(res, "`");
    }
    res
}

fn get_user_args(func: &FuncInfo) -> Vec<(&ParamTypeInfo, &ParamInfo)> {
    func.args
        .iter()
        .filter_map(|v| {
            if let ParamType::Value(ref val) = v.ty {
                Some((v, val))
            } else {
                None
            }
        })
        .collect()
}

fn bracketed_name(arg: &ParamTypeInfo, ty: &ParamInfo) -> String {
    let vcc = ty.ty_info.to_vcc_type();
    let ident = &arg.ident;
    if matches!(ty.kind, ParamKind::Optional) {
        format!("[{vcc} {ident}]")
    } else {
        format!("{vcc} {ident}")
    }
}

/// Print user provided documentation, indenting the `## ...` subsections with a prefix
fn write_docs(docs: &mut String, user_doc: &str, prefix: &str) {
    if !user_doc.is_empty() {
        if user_doc.contains('#') {
            // If the user provided documentation contains a `#`, we need to indent the subsections
            let mut user_doc = user_doc.replace("\n#", &format!("\n#{prefix}"));
            if user_doc.starts_with('#') {
                user_doc.insert_str(0, prefix);
            }
            ln!(docs, "\n{user_doc}");
        } else {
            ln!(docs, "\n{user_doc}");
        }
    }
}
