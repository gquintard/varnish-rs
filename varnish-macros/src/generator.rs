//! The main generator for the varnish vmod.

use std::ffi::CString;
use std::fmt::Write as _;

use proc_macro2::TokenStream;
use quote::quote;
use serde_json::{json, Value};
use sha2::{Digest as _, Sha256};
use syn::{Item, ItemMod, Path};

use crate::gen_func::FuncProcessor;
use crate::gen_objects::ObjProcessor;
use crate::model::VmodInfo;
use crate::names::{ForceCstr, Names, ToIdent};

pub fn render_model(mut item_mod: ItemMod, info: &VmodInfo) -> TokenStream {
    let output = Generator::render(info);
    item_mod
        .content
        .as_mut()
        .unwrap()
        .1
        .insert(0, Item::Verbatim(output));

    quote! { #item_mod }
}

#[derive(Debug, Default)]
pub struct Generator {
    pub names: Names,
    pub file_id: CString,
    pub functions: Vec<FuncProcessor>,
    pub objects: Vec<ObjProcessor>,
}

/// See also <https://varnish-cache.org/docs/7.5/reference/vmod.html>
impl Generator {
    pub fn render(vmod: &VmodInfo) -> TokenStream {
        let mut obj = Self {
            names: Names::new(&vmod.ident),
            file_id: Self::calc_file_id(vmod).force_cstr(),
            ..Self::default()
        };
        for info in &vmod.funcs {
            obj.functions.push(FuncProcessor::from_info(
                obj.names.to_func(info.func_type, &info.ident),
                info,
            ));
        }
        for info in &vmod.objects {
            obj.objects
                .push(ObjProcessor::from_info(obj.names.to_obj(&info.ident), info));
        }
        obj.render_generated_mod(vmod)
    }

    /// Use the entire data model parsed from sources to generate a hash.
    /// Should be somewhat consistent and unique for each set of functions.
    fn calc_file_id(info: &VmodInfo) -> String {
        Sha256::digest(format!("{info:?}").as_bytes())
            .into_iter()
            .fold(String::new(), |mut output, b| {
                let _ = write!(output, "{b:02x}");
                output
            })
    }

    fn gen_priv_structs(tokens: &mut Vec<TokenStream>, name: &str, shared_type: Option<&String>) {
        if let Some(type_name) = shared_type {
            let ident = name.to_ident();
            // The type name is stored as a string, but we already validated we can parse it during the `parse` phase.
            let ty_ident = syn::parse_str::<Path>(type_name).expect("Unable to parse second time");
            let ty_name = type_name.force_cstr();
            // Static methods to clean up the `vmod_priv` object's `T`
            tokens.push(quote! {
               static #ident: vmod_priv_methods = vmod_priv_methods {
                   magic: VMOD_PRIV_METHODS_MAGIC,
                   type_: #ty_name.as_ptr(),
                   fini: Some(vmod_priv::on_fini::<#ty_ident>),
               };
            });
        }
    }

    fn iter_all_funcs(&self) -> impl Iterator<Item = &FuncProcessor> {
        self.functions
            .iter()
            .chain(self.objects.iter().flat_map(|o| o.funcs.iter()))
    }

    fn gen_json(&self) -> String {
        let mod_name = &self.names.mod_name();
        let mut json = Vec::<Value>::new();

        json.push(json! {[
            "$VMOD",
            "1.0",
            mod_name,
            self.names.func_struct_name(),
            self.file_id.to_str().unwrap(),
            // Ohh the irony, this string from VMOD_ABI_Version is the reason
            // why `varnish-sys` must exist. Without it, we could run bindgen
            // from the `varnish` crate directly.  Ohh well.
            //
            // FIXME: figure out a way to assert that the version string used by varnish_macro is the same
            //        as the value accessible by generated code from varnish::ffi::VMOD_ABI_Version.
            //        Currently it seems not possible to do a constant assert at compile time on b-str/c-str equality.
            varnish_sys::ffi::VMOD_ABI_Version.to_str().unwrap(),
            "0",
            "0"
        ]});

        let mut cproto = String::new();
        for obj in &self.objects {
            cproto.push_str(&obj.cproto_typedef_decl);
        }
        for func in self.iter_all_funcs() {
            cproto.push_str(&func.cproto_typedef_decl);
        }
        let _ = write!(cproto, "\nstruct {} {{\n", self.names.func_struct_name());
        for func in self.iter_all_funcs() {
            cproto.push_str(&func.cproto_typedef_init);
        }
        let _ = write!(
            cproto,
            "}};\n\nstatic struct {struct_name} {struct_name};",
            struct_name = self.names.func_struct_name()
        );
        json.push(json! {[ "$CPROTO", cproto ]});

        for func in &self.functions {
            json.push(func.json.clone());
        }

        for obj in &self.objects {
            json.push(obj.json.clone());
        }

        let json = serde_json::to_string_pretty(&json! {json}).unwrap();
        format!("VMOD_JSON_SPEC\u{2}\n{json}\n\u{3}")
    }

    fn render_generated_mod(&self, vmod: &VmodInfo) -> TokenStream {
        let vmod_name_data = self.names.data_struct_name().to_ident();
        let c_name = self.names.mod_name().force_cstr();
        let c_func_name = self.names.func_struct_name().force_cstr();
        let file_id = &self.file_id;
        let mut priv_structs = Vec::new();
        Self::gen_priv_structs(
            &mut priv_structs,
            "PRIV_TASK_METHODS",
            vmod.shared_types.shared_per_task_ty.as_ref(),
        );
        Self::gen_priv_structs(
            &mut priv_structs,
            "PRIV_VCL_METHODS",
            vmod.shared_types.shared_per_vcl_ty.as_ref(),
        );
        let functions = self.iter_all_funcs().map(|f| &f.wrapper_function_body);
        let json = &self.gen_json().force_cstr();
        let export_decls: Vec<_> = self.iter_all_funcs().map(|f| &f.export_decl).collect();
        let export_inits: Vec<_> = self.iter_all_funcs().map(|f| &f.export_init).collect();

        quote!(
            #[allow(
                non_snake_case,
                unused_imports,
                unused_qualifications,
                unused_variables,
            )]
            mod varnish_generated {
                use std::borrow::Cow;
                use std::ffi::{c_char, c_int, c_uint, c_void, CStr};
                use std::ptr::null;
                use varnish::ffi::{
                    VCL_BACKEND,
                    VCL_BOOL,
                    VCL_DURATION,
                    VCL_INT,
                    VCL_IP,
                    VCL_PROBE,
                    VCL_REAL,
                    VCL_STRING,
                    VCL_VOID,
                    VMOD_ABI_Version,
                    VMOD_PRIV_METHODS_MAGIC,
                    vcl_event_e,
                    vmod_priv,
                    vmod_priv_methods,
                    vrt_ctx,
                };
                use varnish::vcl::{Ctx, IntoVCL};
                use super::*;

                #( #priv_structs )*
                #( #functions )*

                #[repr(C)]
                pub struct VmodExports {
                    #(#export_decls,)*
                }

                // #[no_mangle]  // FIXME: no_mangle does not seem to be needed
                pub static VMOD_EXPORTS: VmodExports = VmodExports {
                    #(#export_inits,)*
                };

                #[repr(C)]
                pub struct VmodData {
                    vrt_major: c_uint,
                    vrt_minor: c_uint,
                    file_id: *const c_char,
                    name: *const c_char,
                    func_name: *const c_char,
                    func: *const c_void,
                    func_len: c_int,
                    proto: *const c_char,
                    json: *const c_char,
                    abi: *const c_char,
                }

                unsafe impl Sync for VmodData {}

                // This name must be in the format `Vmod_{modulename}_Data`.
                #[allow(non_upper_case_globals)]
                #[no_mangle]
                pub static #vmod_name_data: VmodData = VmodData {
                    vrt_major: 0,
                    vrt_minor: 0,
                    file_id: #file_id.as_ptr(),
                    name: #c_name.as_ptr(),
                    func_name: #c_func_name.as_ptr(),
                    func_len: size_of::<VmodExports>() as c_int,
                    func: &VMOD_EXPORTS as *const _ as *const c_void,
                    abi: VMOD_ABI_Version.as_ptr(),
                    json: JSON.as_ptr(),
                    proto: null(),
                };

                const JSON: &CStr = #json;
            }
        )
    }
}
