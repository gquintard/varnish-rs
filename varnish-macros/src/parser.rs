//! Parser takes in the token stream and generates the model structure.
//! The model describes the functions, objects, and events that should be exposed to Varnish.
//! The model should be thoroughly validated before generating the output, and is treated as the source of truth.

use darling::ast::NestedMeta;
use darling::FromMeta;
use proc_macro2::TokenStream;
use syn::{Attribute, ImplItem, Item, ItemImpl, ItemMod, Signature, Visibility};

use crate::errors::Errors;
use crate::model::{
    FuncInfo, FuncType, ObjInfo, ParamType, ParamTypeInfo, ReturnTy, ReturnType, SharedTypes,
    VmodInfo, VmodParams,
};
use crate::parser_args::FuncStatus;
use crate::{parser_utils, ProcResult};

pub fn tokens_to_model(args: TokenStream, item_mod: &mut ItemMod) -> ProcResult<VmodInfo> {
    let args = NestedMeta::parse_meta_list(args).map_err(syn::Error::from)?;
    let args = VmodParams::from_list(&args)?;
    let info = VmodInfo::parse(args, item_mod)?;
    Ok(info)
}

impl VmodInfo {
    /// Parse the `mod` item and generate the model of everything
    fn parse(params: VmodParams, item: &mut ItemMod) -> ProcResult<Self> {
        let mut errors = Errors::new();
        let mut funcs = Vec::<FuncInfo>::new();
        let mut objects = Vec::<ObjInfo>::new();
        let mut shared_types = SharedTypes::default();
        let mut has_event = false;
        if let Some((_, content)) = &mut item.content {
            for item in content {
                if let Item::Fn(fn_item) = item {
                    // a function or an event handler
                    let func = FuncInfo::parse(
                        &mut shared_types,
                        &mut fn_item.sig,
                        &fn_item.vis,
                        &mut fn_item.attrs,
                        false,
                    );
                    if let Some(func) = errors.on_err(func) {
                        if matches!(func.func_type, FuncType::Event) {
                            if has_event {
                                errors.add(&fn_item.sig.ident, "Only one event handler is allowed");
                                continue;
                            }
                            has_event = true;
                        }
                        funcs.push(func);
                    }
                } else if let Item::Impl(impl_item) = item {
                    // an object
                    if let Some(obj) = errors.on_err(ObjInfo::parse(impl_item, &mut shared_types)) {
                        objects.push(obj);
                    }
                }
            }
        }
        if funcs.is_empty() && objects.is_empty() {
            errors.add(&item.ident, "No functions or objects found in this module");
        }

        errors.into_result()?;
        Ok(Self {
            params,
            ident: item.ident.to_string(),
            docs: parser_utils::parse_doc_str(&item.attrs),
            shared_types,
            funcs,
            objects,
        })
    }
}

impl ObjInfo {
    /// Parse an `impl` block and treat all public functions as object methods
    fn parse(item_impl: &mut ItemImpl, shared_types: &mut SharedTypes) -> ProcResult<Self> {
        let mut errors = Errors::new();
        let ident = parser_utils::as_simple_ty(item_impl.self_ty.as_ref()).map(|v| v.to_string());

        // Add only one error per object impl declaration
        if item_impl.trait_.as_ref().is_some() {
            errors.add(&item_impl, "Trait impls are not supported for object impls");
        } else if !item_impl.generics.params.is_empty() {
            errors.add(
                &item_impl.generics.params,
                "Generics are not supported for object impls",
            );
        } else if ident.is_none() {
            errors.add(
                &item_impl.self_ty,
                "Expected a simple type for object. If the object is defined elsewhere, use `use` to import it.",
            );
        }

        let mut funcs = Vec::new();
        let mut constructor = None;
        for item in &mut item_impl.items {
            if let ImplItem::Fn(fn_item) = item {
                let Some(func) = errors.on_err(FuncInfo::parse(
                    shared_types,
                    &mut fn_item.sig,
                    &fn_item.vis,
                    &mut fn_item.attrs,
                    true,
                )) else {
                    continue;
                };
                if func.ident == "new" {
                    constructor = Some(func);
                } else {
                    funcs.push(func);
                }
            }
        }

        if constructor.is_none() {
            errors.add(
                &item_impl.self_ty,
                "Object must have a constructor called `new`",
            );
        }

        errors.into_result()?;
        Ok(Self {
            ident: ident.expect("ident err already reported"),
            docs: parser_utils::parse_doc_str(&item_impl.attrs),
            constructor: constructor.expect("ctor err already reported"),
            destructor: FuncInfo {
                func_type: FuncType::Destructor,
                ident: "_fini".to_string(),
                docs: String::new(),
                has_optional_args: false,
                args: Vec::new(),
                returns: ReturnType::Value(ReturnTy::Default),
            },
            funcs,
        })
    }
}

impl FuncInfo {
    /// Parse a function or a method signature
    fn parse(
        shared_types: &mut SharedTypes,
        signature: &mut Signature,
        vis: &Visibility,
        attrs: &mut Vec<Attribute>,
        is_object: bool,
    ) -> ProcResult<Self> {
        let mut errors = Errors::new();

        if !matches!(vis, Visibility::Public(..)) {
            errors.add(
                signature, // cannot use `vis` because it might be `Inherited`
                "Only public functions and impl blocks are allowed inside a `mod` tagged with `#[varnish::vmod]`. Add `pub` or move this function outside of this mod.",
            );
        } else if signature.asyncness.is_some() {
            errors.add(signature, "async functions are not supported");
        } else if signature.unsafety.is_some() {
            errors.add(signature, "unsafe functions are not supported");
        }

        let func_type = if let Some(attr) = parser_utils::remove_attr(attrs, "event") {
            if is_object {
                errors.add(
                    &attr.meta,
                    "Event functions are not supported for object methods",
                );
            }
            FuncType::Event
        } else if is_object {
            if signature.ident == "new" {
                FuncType::Constructor
            } else {
                FuncType::Method
            }
        } else {
            FuncType::Function
        };

        let returns = errors.on_err(ReturnType::parse(&signature.output, func_type));
        let mut status = FuncStatus::new(func_type);
        let mut args = Vec::new();

        for (idx, arg) in signature.inputs.iter_mut().enumerate() {
            let arg = ParamTypeInfo::parse(shared_types, &mut status, idx, arg);
            if let Some(arg) = errors.on_err(arg) {
                args.push(arg);
            }
        }

        let has_optional_args = args
            .iter()
            .any(|arg| matches!(&arg.ty, ParamType::Value(v) if v.is_optional));

        errors.into_result()?;
        Ok(Self {
            func_type,
            ident: signature.ident.to_string(),
            docs: parser_utils::parse_doc_str(attrs),
            has_optional_args,
            returns: returns.expect("returns err already reported"),
            args,
        })
    }
}
