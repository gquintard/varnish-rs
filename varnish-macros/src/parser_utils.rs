use proc_macro2::Ident;
use quote::quote;
use syn::Expr::Lit;
use syn::Lit::Str;
use syn::Meta::NameValue;
use syn::PathArguments::AngleBracketed;
use syn::Type::{Path, Reference};
use syn::{Attribute, ExprLit, GenericArgument, MetaNameValue, PathSegment, Type, TypePath};

use crate::errors::error;
use crate::model::{FuncInfo, ObjInfo};
use crate::ProcResult;

/// iterator to go over all functions in a ObjInfo, including constructor and destructor
pub struct ObjFuncIter<'a> {
    obj: &'a ObjInfo,
    idx: usize,
}

impl<'a> Iterator for ObjFuncIter<'a> {
    type Item = &'a FuncInfo;

    fn next(&mut self) -> Option<Self::Item> {
        let idx = self.idx;
        self.idx += 1;
        match idx {
            0 => Some(&self.obj.constructor),
            1 => Some(&self.obj.destructor),
            idx => self.obj.funcs.get(idx - 2),
        }
    }
}

impl ObjInfo {
    pub fn iter(&self) -> ObjFuncIter {
        ObjFuncIter { obj: self, idx: 0 }
    }
}

/// Remove an attribute from a list of attributes, returning the attribute if found.
pub fn remove_attr(attrs: &mut Vec<Attribute>, name: &str) -> Option<Attribute> {
    attrs
        .iter()
        .position(|attr| attr.path().is_ident(name))
        .map(|idx| attrs.swap_remove(idx))
}

/// Try to get the inner types of the `Result<Ok, Err>` type, or return None if it's not a `Result<Ok, Err>`.
pub fn as_result_type(ty: &Type) -> Option<(&Type, &Type)> {
    if let Path(type_path) = ty {
        if let Some(PathSegment { ident, arguments }) = type_path.path.segments.last() {
            if ident == "Result" {
                if let AngleBracketed(args) = &arguments {
                    if args.args.len() == 2 {
                        if let Some(GenericArgument::Type(ok_ty)) = args.args.first() {
                            if let Some(GenericArgument::Type(err_ty)) = args.args.last() {
                                return Some((ok_ty, err_ty));
                            }
                        }
                    }
                }
            }
        }
    }
    None
}

/// Try to get the inner type of the `Option<T>`, or return None if it's not an `Option<T>`.
pub fn as_option_type(ty: &Type) -> Option<&Type> {
    as_one_gen_type(ty, "Option")
}

/// Try to get the inner type of the `Box<T>`, or return None if it's not a `Box<T>`.
pub fn as_box_type(ty: &Type) -> Option<&Type> {
    as_one_gen_type(ty, "Box")
}

/// Try to get the inner type of `__name__<T>` type with one argument, or return None if it's not a generic type with one argument.
fn as_one_gen_type<'a>(ty: &'a Type, name: &'static str) -> Option<&'a Type> {
    if let Some(GenericArgument::Type(inner_ty)) = as_one_gen_arg(ty, name) {
        Some(inner_ty)
    } else {
        None
    }
}

/// Get the single generic argument of a type, if it's a generic type with one argument.
pub fn as_one_gen_arg<'a>(ty: &'a Type, name: &'static str) -> Option<&'a GenericArgument> {
    if let Path(type_path) = ty {
        if let Some(PathSegment { ident, arguments }) = type_path.path.segments.last() {
            if ident == name {
                if let AngleBracketed(args) = &arguments {
                    if args.args.len() == 1 {
                        return args.args.first();
                    }
                }
            }
        }
    }
    None
}

/// Try to get the inner type of the `&T` reference, or return None if it's not a `&T` reference.
pub fn as_ref_ty(ty: &Type) -> Option<&Type> {
    if let Reference(rf) = ty {
        if rf.mutability.is_none() {
            return Some(&rf.elem);
        }
    }
    None
}

/// Try to get the inner type of the `&mut T` reference, or return None if it's not a `&mut T` reference.
pub fn as_ref_mut_ty(ty: &Type) -> Option<&Type> {
    if let Reference(rf) = ty {
        if rf.mutability.is_some() {
            return Some(&rf.elem);
        }
    }
    None
}

/// Try to get the ident of a simple type, or return None if it's not a simple type.
pub fn as_simple_ty(ty: &Type) -> Option<&Ident> {
    if let Path(TypePath { qself: None, path }) = ty {
        path.get_ident()
    } else {
        None
    }
}

/// Save/validate shared mut `T` into the store. Must be declared as `&mut Option<Box<T>>`
pub fn parse_shared_mut(store: &mut Option<String>, arg_ty: &Type) -> ProcResult<()> {
    let val = as_ref_mut_ty(arg_ty)
        .and_then(as_option_type)
        .and_then(as_box_type);
    store_shared(store, arg_ty, val, true)
}

/// Save/validate shared ref `T` into the store. Must be declared as `Option<&T>`
pub fn parse_shared_ref(store: &mut Option<String>, arg_ty: &Type) -> ProcResult<()> {
    let val = as_option_type(arg_ty).and_then(as_ref_ty);
    store_shared(store, arg_ty, val, false)
}

/// When processing a fn arg tagged with `#[shared_per_task]` or `#[shared_per_vcl]`,
/// we need to ensure that the shared type is the same everywhere. This function
/// stores the shared type into the `store`, or if it is already non-None, it checks
/// that the type is the same.  This is a helper function for `parse_shared_mut` and `parse_shared_ref`.
fn store_shared(
    store: &mut Option<String>,
    arg_ty: &Type,
    ty: Option<&Type>,
    is_mut: bool,
) -> ProcResult<()> {
    let Some(ty) = ty.and_then(|ty| {
        if let Path(TypePath { qself: None, path }) = ty {
            Some(quote! { #path }.to_string())
        } else {
            None
        }
    }) else {
        let msg = if is_mut {
            "This params must be declared as `&mut Option<Box<...>>`"
        } else {
            "This params must be declared as `Option<&...>`"
        };
        Err(error(arg_ty, msg))?
    };

    if let Some(t) = store {
        if t != &ty {
            let msg = format!(
                "Shared type must be the same everywhere. Another shared param used type `{t}`."
            );
            Err(error(arg_ty, &msg))?;
        }
    } else {
        // Ensure we can parse the types later, but we need to store it as a string to avoid lifetime issues
        if let Err(e) = syn::parse_str::<syn::Path>(&ty) {
            let msg = format!(
                "Internal error, please report: unable to re-parse this from a string '{ty}'"
            );
            Err(syn::Error::new(e.span(), msg))?;
        }

        *store = Some(ty);
    }

    Ok(())
}

/// Parse the doc string from the `#[doc]` attributes, and return it as a string.
pub fn parse_doc_str(attrs: &[Attribute]) -> String {
    // Adapted from https://github.com/hasura/graphql-engine/blob/a29101a3e9fe8c624fb09ec892c21da4b2bdaaba/v3/crates/utils/opendds-derive/src/helpers.rs#L35
    // Under MIT / Apache2.0 license
    let attrs = attrs
        .iter()
        .filter_map(|attr| {
            if attr.path().is_ident("doc") {
                if let NameValue(MetaNameValue {
                    value:
                        Lit(ExprLit {
                            lit: Str(ref lit_str),
                            ..
                        }),
                    ..
                }) = attr.meta
                {
                    Some(lit_str.value())
                } else {
                    None
                }
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    let mut lines = attrs
        .iter()
        .flat_map(|a| a.split('\n'))
        .map(str::trim)
        .skip_while(|s| s.is_empty())
        .collect::<Vec<_>>();

    if let Some(&"") = lines.last() {
        lines.pop();
    }

    // Added for backward-compatibility, but perhaps we shouldn't do this
    // https://github.com/rust-lang/rust/issues/32088
    if lines.iter().all(|l| l.starts_with('*')) {
        for line in &mut lines {
            *line = line[1..].trim();
        }
        while let Some(&"") = lines.first() {
            lines.remove(0);
        }
    }

    lines.join("\n")
}
