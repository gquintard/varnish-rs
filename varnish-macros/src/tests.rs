use std::fs::read_to_string;
use std::path::PathBuf;

use insta::{assert_snapshot, with_settings};
use proc_macro2::TokenStream;
use quote::quote;
use regex::Regex;
use syn::ItemMod;

use crate::gen_docs::gen_doc_content;
use crate::generator::render_model;
use crate::parser::tokens_to_model;
use crate::parser_utils::remove_attr;

/// Read content of the ../tests/pass directory that should also pass full compilation tests,
/// parse them and create snapshots of the model and the generated code.
#[test]
fn parse_passing_tests() {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("tests/pass/*.rs");
    for file in glob::glob(path.to_str().unwrap()).unwrap() {
        let filepath = file.unwrap();
        let file = filepath.to_str().unwrap();
        eprintln!("Processing file: {file}");
        let src = read_to_string(&filepath).expect("unable to read file");
        let syntax = syn::parse_file(&src).expect("unable to parse file");
        let mut has_vmod = false;
        for item in syntax.items {
            if let syn::Item::Mod(mut item) = item {
                assert!(!has_vmod, "Multiple vmod modules found in file {file}");
                has_vmod = true;
                // FIXME: use this attribute as an arg for the test
                let _arg = remove_attr(&mut item.attrs, "vmod").unwrap();
                let name = format!(
                    "{}_{}",
                    filepath.file_stem().unwrap().to_string_lossy(),
                    item.ident
                );
                test(&name, quote! {}, item);
                // FIXME: pass proper attribute info
                // test(&name, quote! { #arg }, item);
            }
        }
        assert!(has_vmod, "No vmod modules found in file {file}");
    }
}

fn test(name: &str, args: TokenStream, mut item_mod: ItemMod) {
    // panic!("{name} {args:?} {item_mod:?}");
    with_settings!({ snapshot_path => "../../varnish/snapshots", omit_expression => true, prepend_module_to_snapshot => false }, {
        let Ok(info) = tokens_to_model(args, &mut item_mod).map_err(|err| {
            // On error, save the error output as a snapshot and return early.
            let err = err.into_compile_error();
            with_settings!({ snapshot_suffix => "error" }, { assert_snapshot!(name, err) });
        }) else { return };

        with_settings!({ snapshot_suffix => "model" }, { assert_snapshot!(name, format!("{info:#?}")) });
        with_settings!({ snapshot_suffix => "docs" }, { assert_snapshot!(name, gen_doc_content(&info)) });

        let file = render_model(item_mod, &info).to_string();
        let parsed = match syn::parse_file(&file) {
            Ok(v) => v,
            Err(e) => {
                // We still save the generated code, but fail the test. Allows easier error debugging.
                with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, file) });
                panic!("Failed to parse generated code in test {name}: {e}");
            }
        };

        let generated = prettyplease::unparse(&parsed);

        // Use regex to remove "Varnish 7.5.0 eef25264e5ca5f96a77129308edb83ccf84cb1b1" and similar.
        // Also removes any pre-builds and other versions because we assume a double-quote at the end.
        // TODO: Once MSRV is higher than 1.80, use static LazyLock<Regex>
        let re_varnish_ver = Regex::new(r"Varnish \d+\.[-+. 0-9a-z]+").unwrap();
        let code = re_varnish_ver.replace_all(&generated, "Varnish (version) (hash)");
        with_settings!({ snapshot_suffix => "code" }, { assert_snapshot!(name, code) });

        // Extract JSON string
        let pat = "const JSON: &CStr = c\"";
        let json = if let Some(pos) = code.find(pat) {
            let json = &code[pos + pat.len()..];
            json.split("\";\n").next()
        } else {
            None
        };

        let json = &json
            .unwrap_or("")
            .replace("\\\"", "\"")
            .replace("\\u{2}", "\u{2}")
            .replace("\\u{3}", "\u{3}")
            .replace("\\\\", "\\")
            // this is a bit of a hack because the double-escaping gets somewhat incorrectly parsed
            .replace("\\n", "\n");

        with_settings!({ snapshot_suffix => "json" }, { assert_snapshot!(name, json) });
    });
}
