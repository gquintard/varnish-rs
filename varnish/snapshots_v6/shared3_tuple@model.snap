---
source: varnish-macros/src/tests.rs
---
VmodInfo {
    params: VmodParams {
        docs: None,
    },
    ident: "tuple",
    docs: "",
    funcs: [
        FuncInfo {
            func_type: Function,
            ident: "ref_to_slice_lifetime",
            docs: "",
            has_optional_args: false,
            args: [
                ParamTypeInfo {
                    ident: "tsk_vals",
                    docs: "",
                    ty: SharedPerTask,
                },
            ],
            output_ty: Bytes,
            out_result: false,
        },
    ],
    objects: [],
    shared_types: SharedTypes {
        shared_per_task_ty: Some(
            "PerTask < '_ >",
        ),
        shared_per_vcl_ty: None,
    },
}
